sig-azure/0018-20180711-azure-availability-zones.md

```diff
@@ -0,0 +1,238 @@
+
+## Summary
+
+This proposal aims to add [Azure Availability Zones (AZ)](https://azure.microsoft.com/en-us/global-infrastructure/availability-zones/) support to Kubernetes.
+
+## Scopes and Non-scopes
+
+### Scopes
+
+The proposal includes required changes to support availability zones for various functions in Azure cloud provider and AzureDisk volumes:
+
+- Detect availability zones automatically when registering new nodes and node's label `failure-domain.beta.kubernetes.io/zone` will be replaced with AZ instead of fault domain
+- LoadBalancer and PublicIP will be provisioned with zone redundant
+- `GetLabelsForVolume` interface will be implemented for Azure managed disks and it will also be added to `PersistentVolumeLabel` admission controller so as to support DynamicProvisioningScheduling
+
+### Non-scopes
+
+Provisioning Kubernetes masters and nodes with availability zone support is not included in this proposal. It should be done in the provisioning tools (e.g. acs-engine). Azure cloud provider will auto-detect the node's availability zone if `availabilityZones` option is configured for the Azure cloud provider.
+
+## AZ label format
+
+Currently, Azure nodes are registered with label `failure-domain.beta.kubernetes.io/zone=faultDomain`.
+
+The format of fault domain is numbers (e.g. `1` or `2`), which is in same format with AZ (e.g. `1` or `3`). If AZ is using same format with faultDomain, then there'll be scheduler issues for clusters with both AZ and non-AZ nodes. So AZ will use a different format in kubernetes: `<region>-<AZ>`, e.g. `centralus-1`.
+
+The AZ label will be applied in multiple Kubernetes resources, e.g.
+
+- Nodes
+- AzureDisk PersistentVolumes
+- AzureDisk StorageClass
+
+## Cloud provider options
+
+Because only standard load balancer is supported with AZ, it is a prerequisite to enable AZ for the cluster.
+
+Standard load balancer has been added in Kubernetes v1.11, related options include:
+
+| Option                      | Default | **AZ Value**  | Releases | Notes                                 |
+| --------------------------- | ------- | ------------- | -------- | ------------------------------------- |
+| loadBalancerSku             | basic   | **standard**  | v1.11    | Enable standard LB                    |
+| excludeMasterFromStandardLB | true    | true or false | v1.11    | Exclude master nodes from LB backends |
+
+These options should be configured in Azure cloud provider configure file (e.g. `/etc/kubernetes/azure.json`):
+
+```json
+{
+    ...,
+    "loadBalancerSku": "standard",
+    "excludeMasterFromStandardLB": true
+}
+```
+
+Note that with standard SKU LoadBalancer, `primaryAvailabitySetName` and `primaryScaleSetName` is not required because all available nodes (with configurable masters via `excludeMasterFromStandardLB`) are added to LoadBalancer backend pools.
+
+## Node registration
+
+When nodes are started, kubelet automatically adds labels to them with region and zone information:
```

Comment:
allowing nodes to self-label has been problematic (enumerated in https://github.com/kubernetes/community/pull/911). is there a reason we wouldn't have the node controller pull this info from the cloud provider and manage the labels on the node objects?

Reply:
cc @mikedanese

Reply:
there're actually two ways supported today:

- Without cloud-controller-manager (kubelet --cloud-privider=azure), [kubelet registers itself](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L320) with zone labels
- With cloud-controller-manager (kubelet --cloud-provider=external), [node controller](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L391) adds zone lables

Until cloud-controller-manager GA (probably v1.13 or v1.14), both way should be supported.

Reply:
> Until cloud-controller-manager GA (probably v1.13 or v1.14), both way should be supported

I agree we need to support the old way until we can transition away from it, but I don't think we should expand it to new labels if possible

Reply:
> I agree we need to support the old way until we can transition away from it, but I don't think we should expand it to new labels if possible

Agreed. this is only for existing zone labels.

Reply:
> Agreed. this is only for existing zone labels.

great, thanks for confirming